src/features/diagnosticsProvider.ts

```diff
@@ -124,47 +127,56 @@ class DiagnosticsProvider extends AbstractSupport {
 
     private _validationAdvisor: Advisor;
     private _disposable: CompositeDisposable;
-    private _documentValidations: { [uri: string]: vscode.CancellationTokenSource } = Object.create(null);
-    private _projectValidation: vscode.CancellationTokenSource;
     private _diagnostics: vscode.DiagnosticCollection;
+    private _validateCurrentDocumentPipe = new Subject<vscode.TextDocument>();
+    private _validateAllPipe = new Subject();
+    private _analyzersEnabled: boolean;
+    private _subscriptions: Subscription[] = [];
     private _suppressHiddenDiagnostics: boolean;
 
     constructor(server: OmniSharpServer, validationAdvisor: Advisor) {
         super(server);
 
+        this._analyzersEnabled = vscode.workspace.getConfiguration('omnisharp').get('enableRoslynAnalyzers', false);
         this._validationAdvisor = validationAdvisor;
         this._diagnostics = vscode.languages.createDiagnosticCollection('csharp');
         this._suppressHiddenDiagnostics = vscode.workspace.getConfiguration('csharp').get('suppressHiddenDiagnostics', true);
 
-        let d1 = this._server.onPackageRestore(this._validateProject, this);
-        let d2 = this._server.onProjectChange(this._validateProject, this);
-        let d4 = vscode.workspace.onDidOpenTextDocument(event => this._onDocumentAddOrChange(event), this);
-        let d3 = vscode.workspace.onDidChangeTextDocument(event => this._onDocumentAddOrChange(event.document), this);
-        let d5 = vscode.workspace.onDidCloseTextDocument(this._onDocumentRemove, this);
-        let d6 = vscode.window.onDidChangeActiveTextEditor(event => this._onDidChangeActiveTextEditor(event), this);
-        let d7 = vscode.window.onDidChangeWindowState(event => this._OnDidChangeWindowState(event), this);
-        this._disposable = new CompositeDisposable(this._diagnostics, d1, d2, d3, d4, d5, d6, d7);
-
-        // Go ahead and check for diagnostics in the currently visible editors.
-        for (let editor of vscode.window.visibleTextEditors) {
-            let document = editor.document;
-            if (this.shouldIgnoreDocument(document)) {
-                continue;
-            }
-
-            this._validateDocument(document);
-        }
+        this._subscriptions.push(this._validateCurrentDocumentPipe
+            .asObservable()
+            .pipe(throttleTime(750))
+            .subscribe(async x => await this._validateDocument(x)));
+
+        this._subscriptions.push(this._validateAllPipe
+            .asObservable()
+            .pipe(throttleTime(3000))
+            .subscribe(async () => {
+                try {
+                    if (this._validationAdvisor.shouldValidateAll()) {
+                        await this._validateSmallWorkspace();
+                    }
+                    else if (this._validationAdvisor.shouldValidateFiles()) {
+                        await this._validateLargeWorkspace();
+                    }
+                } catch { }
```

Comment:
Why swallowing all exceptions here?

Reply:
I think this moved from original version, i don't see any reason for it in current form -> remove.